criu/pages-compress.c

```diff
@@ -0,0 +1,208 @@
+#include <limits.h>
+#include <pthread.h>
+#include <sched.h>
+#include <stdio.h>
+#include <sys/resource.h>
+#include <unistd.h>
+
+#include <compel/plugins/std/syscall-codes.h>
+
+#include "pages-compress.h"
+
+#include "crtools.h"
+#include "image.h"
+#include "log.h"
+#include "memfd.h"
+#include "sched.h"
+#include "util.h"
+
+#include <lz4/programs/lz4io.h>
+#include <lz4/lib/lz4frame.h>
+
+#undef LOG_PREFIX
+#define LOG_PREFIX "decompress: "
+
+const char *temp_path_decompressed = "/tmp/decompressed";
+
+int compress_images(void)
+{
+	char format[PATH_MAX];
+	char pathsrc[PATH_MAX];
+	char pathdst[PATH_MAX];
+	const long pid = getpid();
+	const int dfd = get_service_fd(IMG_FD_OFF);
+	const int id = 1; // the only image we'd like to compress
+	int ret;
+
+	LZ4IO_prefs_t *lz4_prefs = LZ4IO_defaultPreferences();
+	LZ4IO_setBlockSize(lz4_prefs, 128 * 1024);
+
+	pr_debug("Compressing pages\n");
+
+	sprintf(format, "%s/%s", "/proc/%ld/fd/%d", imgset_template[CR_FD_PAGES].fmt);
+	sprintf(pathsrc, format, pid, dfd, id);
+
+	sprintf(format, "%s/%s", "/proc/%ld/fd/%d", imgset_template[CR_FD_PAGES_COMP].fmt);
+	sprintf(pathdst, format, pid, dfd, id);
+
+	ret = LZ4IO_compressFilename(pathsrc, pathdst, 8, lz4_prefs);
+	LZ4IO_freePreferences(lz4_prefs);
+
+	if (ret) {
+		pr_err("Can't compress %s to %s\n", pathsrc, pathdst);
+		return -1;
+	}
+	if (unlink(pathsrc)) {
+		pr_perror("Can't delete uncompressed source pages image: %s\n", pathsrc);
+		return -1;
+	}
+	return 0;
+}
+
+static int decompress_image(int comp_fd) {
+
+	const int compbufsize = 64 * 1024;
+	char compbuf[compbufsize];
```

Comment:
Couldn't you rather `mmap()` and `madvise()` the whole file?

Reply:
I think this is a good idea, however now I'm not sure about the max size of a compressed image, and if we can use memory blocks of any size while decompressing, or need to limit it.

Reply:
On 64-bit platforms there is really no limit.
On 32-bit platforms there may be some problem if the image size is getting close to the 2GB (or 3GB) address space limit but I do not think we need to care about it, do we? @AntonKozlov?

Reply:
We don't pay a lot of attention to 32 bit. And since we are going anyway to mmap the image, there is a dependency on being able to mmap the image. Mmaping compressed image will doulbe the mmaping size (at most), so that does not look critical. So the proposal makes sense to me, without considering efforts to implement that.